src/libraries/Common/src/System/Net/Http/aspnetcore/Http3/QPack/QPackDecoder.cs

```diff
@@ -105,64 +103,53 @@ private enum State
         private const int StringLengthPrefix = 7;
         private const byte HuffmanMask = 0x80;
 
-        private const int DefaultStringBufferSize = 64;
+        private const int HeaderStaticIndexUnset = -1; // Static index starts at 0
 
         private readonly int _maxHeadersLength;
         private State _state = State.RequiredInsertCount;
 
-        private byte[] _stringOctets;
-        private byte[] _headerNameOctets;
-        private byte[] _headerValueOctets;
-
         // s is used for whatever s is in each field. This has multiple definition
         private bool _huffman;
-        private int? _index;
 
         private byte[]? _headerName;
+        private int _headerStaticIndex;
         private int _headerNameLength;
         private int _headerValueLength;
         private int _stringLength;
         private int _stringIndex;
+
         private IntegerDecoder _integerDecoder;
+        private byte[]? _stringOctets;
+        private byte[]? _headerNameOctets;
+        private byte[]? _headerValueOctets;
+        private (int start, int length)? _headerNameRange;
+        private (int start, int length)? _headerValueRange;
 
         private static ArrayPool<byte> Pool => ArrayPool<byte>.Shared;
 
-        private static void ReturnAndGetNewPooledArray(ref byte[] buffer, int newSize)
-        {
-            byte[] old = buffer;
-            buffer = null!;
-
-            Pool.Return(old, clearArray: true);
-            buffer = Pool.Rent(newSize);
-        }
-
         public QPackDecoder(int maxHeadersLength)
         {
             _maxHeadersLength = maxHeadersLength;
-
-            // TODO: make allocation lazy? with static entries it's possible no buffers will be needed.
-            _stringOctets = Pool.Rent(DefaultStringBufferSize);
-            _headerNameOctets = Pool.Rent(DefaultStringBufferSize);
-            _headerValueOctets = Pool.Rent(DefaultStringBufferSize);
+            _headerStaticIndex = HeaderStaticIndexUnset;
         }
 
         public void Dispose()
         {
             if (_stringOctets != null)
             {
-                Pool.Return(_stringOctets, true);
+                Pool.Return(_stringOctets);
```

Comment:
I'm not sure why these arrays are being cleared.

Example of similar usage of an array that isn't being cleared:
https://github.com/dotnet/runtime/blob/bddd36907ad276d3601f3eeed0936af2b76ef01c/src/libraries/Common/src/System/Text/ValueStringBuilder.cs#L310-L319

Reply:
We generally don't clear when returning to the array pool.  Main exception to that is when we know the data is likely to be extra sensitive, e.g. in crypto.  Seems fine to not clear here.  We're not elsewhere in HttpClient.